gtfs_rt_validator_api.py

```diff
@@ -25,42 +72,169 @@ def validate(gtfs_file, rt_path, out_file=None, verbose=False):
-            fs.glob(f"{gtfs_rt_date}T*/")
+        # TODO: remove hard-coding
+        if gtfs_rt_glob_path is None:
+            raise ValueError("One of gtfs rt glob path or date must be specified")
+
+
+        download_rt_files(dst_path_rt, fs, glob_path = gtfs_rt_glob_path)
+
+        print("Validating data")
+        validate(f"{dst_path_gtfs}.zip", dst_path_rt, verbose=verbose)
+
+        if results_bucket:
+            # validator stores results as {filename}.results.json
+            print(f"Putting data into results bucket: {results_bucket}")
+
+            # fetch all results files created by the validator
+            all_results = list(Path(dst_path_rt).glob("*.results.json"))
+
+            
+            final_json_dir.mkdir(exist_ok=True)
+            final_files = []
+            for result in all_results:
+                # we appended a final timestamp to the files so that the validator
+                # can use it to order them during validation. here, we remove that
+                # timestamp, so we can use a single wildcard to select, eg..
+                # *trip_updates.results.json
+                result_out = "__".join(result.name.split("__")[:-1])
+
+                json_to_newline_delimited(result, final_json_dir / result_out)
+                final_files.append(final_json_dir / result_out)
+            
+
+            fs.put(final_files, results_bucket)
+
+    except Exception as e:
+        if isinstance(tmp_dir, TemporaryDirectory):
+            tmp_dir.cleanup()
+
+        raise e
+
+
+def download_gtfs_schedule_zip(gtfs_schedule_path, dst_path, fs):
+    # fetch and zip gtfs schedule
+    fs.get(gtfs_schedule_path, dst_path, recursive=True)
+    shutil.make_archive(dst_path, "zip", dst_path)
+    
+
+def download_rt_files(dst_dir, fs=None, date="2021-08-01", glob_path=None):
+    """Download all files for an GTFS RT feed (or multiple feeds)
+
+    If date is specified, downloads daily data for all feeds. Otherwise, if
+    glob_path is specified, downloads data for a single feed.
+
+    Parameters:
+        date: date of desired feeds to download data from (e.g. 2021-09-01)
+        glob_path: if specified, the path (including a wildcard) for downloading a 
+                   single feed.
+
+    """
+    if fs is None:
+        raise NotImplementedError("Must specify fs")
+
+    # {date}T{timestamp}/{itp_id}/{url_number}
+    all_files = fs.glob(glob_path) if glob_path else fs.glob(f"{RT_BUCKET_FOLDER}/{date}*/*/*/*")
+
+    to_copy = []
+    out_feeds = defaultdict(lambda: [])
+    for src_path in all_files:
+        dt, itp_id, url_number, src_fname = src_path.split("/")[-4:]
+        if glob_path:
+            dst_parent = Path(dst_dir)
+        else:
+            # if we are downloading multiple feeds, make each feed a subdir
+            dst_parent = Path(dst_dir) / itp_id / url_number
+
+        dst_parent.mkdir(parents=True, exist_ok=True)
+
+        # TODO: check that this works
```

Comment:
Does it work?

Reply:
whoops--yes, It's run in the integration tests!